### DISCUSSION TITLE
Handling structured points

### DISCUSSION THREAD

**Discussion by bhawesh-bks** [pyvista/pyvista#4401] [Q&A]
So i have a dataset which consits of x,y,z and a target column. I want to plot a 3d grid plot but the grid is not exactly how i want it to be. It is distorted. Please note that there are some null values in the target column which can be filled by 0. Below is the code that I have used and I have also attached the distorted 3 d plot. Please correct my code.

```py
df=pd.read_csv('final_data.csv')
df=df.drop(['Unnamed: 0'],axis=1)
df['target'] = df['target'].fillna(0)
x=len(np.unique(df['x']))
y=len(np.unique(df['y']))
z=len(df)/(x*y)
print(x,y,z)
coords = df[['x', 'y', 'z']].values
pts = pv.PolyData(coords)
pts = df[['x', 'y', 'z']].values
dt = df['target'].values
poly = pv.PolyData(pts)
poly['target'] = dt
grid = pv.StructuredGrid()
grid.points = pts
grid['target'] = dt
grid.dimensions = (61, 57, 37)
grid.plot(notebook=0,show_grid=True,show_edges=True)
```
![Screenshot (3)](https://github.com/pyvista/pyvista/assets/97894832/1d5ac2e9-3d8f-44f1-9fd9-de853d356166)
[final_data (2).csv](https://github.com/pyvista/pyvista/files/11436098/final_data.2.csv)

**Comment by banesullivan** [comment]
The DataFrame/CSV seems malformed to me. Your data values do not match their coordinates row-wise. I noticed this because it's a common problem I've seen repeatedly with structured points. To align them, you need to properly ravel/unravel the arrays (I suspect you just ran `arr.flatten()` for your structure coodinates and the ordering got messed up). This is a common C vs F ordering issue.

Otherwise, this is how you can plot as both a point cloud and a `StructuredGrid`. Some notes:

- Make sure to cast the coordinates as floats (PyVista handles these better than BigInt)
- The dimensions of the grid are the internal structure/raveling of the array, not XYZ in that order. So this is `(nz, ny, nx)` for your example.


```py
import pandas as pd
import pyvista as pv
import numpy as np

df = pd.read_csv('final_data.2.csv')
df = df.drop(['Unnamed: 0'], axis=1)
df['target'] = df['target'].fillna(0)

# Cast coordinates as float
coords = df[['x', 'y', 'z']].values.astype(float)

# Find dimensionality of structured points 
nx = len(np.unique(df['x']))
ny = len(np.unique(df['y']))
nz = len(np.unique(df['z']))

# Properly unravel data values to match coordinate ordering
dt = df['target'].values.reshape((nx, ny, nz, 1), order='F').reshape((-1,), order='C')
```

```py
# Create point cloud
poly = pv.PolyData(coords)
poly.point_data['target'] = dt
poly.plot()
```

![screenshot (1)](https://github.com/pyvista/pyvista/assets/22067021/8f411183-666d-4081-baa0-c42e694e9812)



```py
# Create StructuredGrid
grid = pv.StructuredGrid()
grid.points = coords  # These need to be floats
grid.dimensions = (nz, ny, nx)
grid.point_data['target'] = dt

grid.plot(show_edges=True)
``` 
![screenshot (2)](https://github.com/pyvista/pyvista/assets/22067021/2cf60fed-2341-47ee-b455-383d6a938e67)


```py
grid.contour().plot()
```
![screenshot (3)](https://github.com/pyvista/pyvista/assets/22067021/5714d7d9-d700-4c46-9ad6-187fa1d23348)
